icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -9,39 +16,404 @@ namespace icicle {
+        const uint64_t sub_tree_leaves_offset = (element_offset / sub_tree_leaves_size) * sub_tree_leaves_size;
+        const std::byte* sub_tree_leaves = &(leaves[sub_tree_leaves_offset]);
+        sub_tree.build(sub_tree_leaves, sub_tree_leaves_size, config);
+
+        // retrieve from the sub tree the path and increment path
+        const uint64_t sub_tree_input_chunk_offset = element_offset % sub_tree_leaves_size;
+        sub_tree.copy_to_path_from_store_min_layer(sub_tree_input_chunk_offset, is_pruned, path);
+      }
+
+      copy_to_path_from_store_min_layer(input_chunk_offset, is_pruned, path);
+      // print_proof(merkle_proof);
+      return eIcicleError::SUCCESS;
+    }
+
+    // Debug functions
+    eIcicleError print_tree(const std::byte* leaves, uint64_t leaves_size) const
+    {
+      ICICLE_ASSERT(m_tree_already_built) << "Tree cannot print before built";
+      std::cout << "Print tree:" << std::endl;
+      std::cout << "Leaves: size=" << m_leaf_element_size << std::endl;
+      print_bytes(leaves, leaves_size / m_leaf_element_size, m_leaf_element_size);
+      for (int layer_idx = m_output_store_min_layer; layer_idx < m_layers.size(); ++layer_idx) {
+        std::cout << std::dec << "Layer " << layer_idx << ": " << m_layers[layer_idx].m_hash.input_default_chunk_size()
+                  << " -> " << m_layers[layer_idx].m_hash.output_size() << std::endl;
+        print_bytes(
+          m_layers[layer_idx].m_results.data(), m_layers[layer_idx].m_nof_hashes,
+          m_layers[layer_idx].m_hash.output_size());
+      }
+      return eIcicleError::SUCCESS;
+    }
+
+    eIcicleError print_proof(const MerkleProof& proof) const
+    {
+      const auto [leaf_data, leaf_size, leaf_index] = proof.get_leaf();
+      std::cout << "Print Proof: is_pruned(" << proof.is_pruned() << "), leaf_idx(" << leaf_index << ")" << std::endl;
+      std::cout << "Leaves:" << std::endl;
+      print_bytes(leaf_data, leaf_size / m_leaf_element_size, m_leaf_element_size);
+
+      auto [path, path_size] = proof.get_path();
+      for (uint layer_idx = 0; layer_idx < m_layers.size() - 1; ++layer_idx) {
+        std::cout << "Layer " << layer_idx << " outputs:" << std::endl;
+        const uint output_size = m_layers[layer_idx].m_hash.output_size();
+        const uint nof_hashes =
+          m_layers[layer_idx].m_nof_hashes / m_layers[layer_idx + 1].m_nof_hashes - (proof.is_pruned() ? 1 : 0);
+
+        print_bytes(path, nof_hashes, output_size);
+        path += nof_hashes * output_size;
+      }
+      auto [root, root_size] = proof.get_root();
+      std::cout << "Layer " << m_layers.size() - 1 << "(Root):" << std::endl;
+      print_bytes(root, 1, root_size);
+      return eIcicleError::SUCCESS;
+    }
+
+    eIcicleError get_hash_result(int layer_index, int hash_index, const std::byte*& hash_result /*OUT*/) const
+    {
+      ICICLE_ASSERT(m_tree_already_built) << "Tree cannot get hash result before built";
```

Comment:
if you anyway assert or return success then just return the pointer or null with an ICICLE_LOG_ERROR.
this reference to pointer is less clear when reading the top level only

Reply:
fixed